feat(sltp): base ui for sltp edit modal

moo-onthelawn opened the pull request:
<!-- Featured screenshots/recordings -->
This PR:
- Creates the skeleton UI for the SL/TP edit modal as highlighted by mocks [here](https://www.figma.com/file/fMQodZKzn9B5aZTN5G0fLJ/dYdX-%E2%80%BA-Desktop?type=design&node-id=24635-71787&mode=design&t=yLCLgluJQmOcqVlb-0)
    -  Handles navigation when clicking "View Orders"
    - Adds tooltips
- Noted followups
    - a lot of the values will be filled out by abacus (which is still a WIP, so for now they are either empty / hardcoded); left TODOs to signify 
    - the size slider doesn't match the mocks (it's very much just the basics rn), will be polished up in later PRs
    - I haven't added the Profit/Loss [labels](https://www.figma.com/file/fMQodZKzn9B5aZTN5G0fLJ/dYdX-%E2%80%BA-Desktop?type=design&node-id=24635-72442&mode=design&t=yLCLgluJQmOcqVlb-0) next to Take Profit/Stop Loss in the dialog; want to clarify some things design wise before doing that (CT-682)
    - Add trigger(s) onClick doesn't do anything yet

<!-- Overall purpose of the PR -->
 ### Stories 
| Icon Trigger | Icon Trigger - Expanded | Slot Trigger | Slot Trigger - Expanded | Order Size Slider |
| ----- | ----- |  ----- |  ----- | ----- | 
| <img width="1260" alt="Screenshot 2024-03-13 at 2 50 20 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/b9b77d0b-d9ea-4f3e-a654-3a029624cd8b"> | <img width="1257" alt="Screenshot 2024-03-13 at 2 50 55 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/275699f8-b0c6-40c8-9e26-5d3b72f0228d"> | <img width="1248" alt="Screenshot 2024-03-13 at 2 50 42 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/c7e7a2ec-1074-4590-b8ad-4ff418d448e0"> | <img width="1260" alt="Screenshot 2024-03-13 at 2 50 50 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/80e3c714-1f12-4e85-93cd-d21ef767b4a6"> | <img width="1487" alt="Screenshot 2024-03-15 at 4 00 24 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/d31b3dba-ee5d-482b-b402-c70ac8ff2690"> |

### Dialog
| Dialog | Dialog - Expanded States | Dialog - multiOrders | Dialog - multiOrders on both |
 | ----- | ----- |  ----- | ----- | 
| <img width="531" alt="Screenshot 2024-03-13 at 2 54 30 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/cda49131-b510-4449-bea4-92103d601dd0"> | <img width="519" alt="Screenshot 2024-03-13 at 2 54 52 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/f9c8f043-ff0c-47b8-9ed0-a88d1323f7fe"> | <img width="528" alt="Screenshot 2024-03-13 at 2 59 39 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/b1003bdc-4cb6-41ec-8daa-5c1b612d4463"> | <img width="540" alt="Screenshot 2024-03-20 at 3 17 45 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/81dd75b3-67b3-425c-ba77-bc93984b23bf"> |



https://github.com/dydxprotocol/v4-web/assets/70078372/b066e9bb-cef5-4df2-9028-66de0371dbb1

---




<!-- Reorder/delete the following sections accordingly: -->

## Views

* New: 
```
TriggersDialog

forms/TriggersForm
    AdvancedTriggersOptions
    LimitPriceInputs
        // Used for "Limit Price", not currently functional, will be filled out with abacus work in CT-625
        // Collapsible (only renders contents when checked)
    OrderSizeInput
        // Used for "Custom Amount", not currently functional, will be filled out with abacus work in CT-625
        // Collapsible (only renders contents when checked)
    OrderSizeSlider
        // Used for "Custom Amount", currently a skeleton (minimal styling), will polish off in follow up PR
    TriggerOrderInputs
        // Used to render SL/TP orders if they exist (or a multiple orders button).
        // Does currently render the multiple orders button when there are > 1 SL/TP, but otherwise not functional, will be filled out with abacus work in CT-625
    TriggersForm
        // Composes TriggersForm inputs 
```

* `views/tables/PositionsTable/PositionsActionsCell`
  * hook into calling into `TriggersDialog`

## Components

* `Collapsible`
  * Updated to allow for a `slotTrigger` (instead of existing `IconTrigger`)
    * Notable diff between the two is that the collapsible state will only be updated by clicking the `slotTrigger` (instead of the `icon` + `label` as with the `IconTrigger` case)
  * Updated corresponding stories file

## Constants/Types

* `constants/dialogs`
  * Add new `Triggers` dialog type
* `constants/tooltips/index.ts` + `constants/tooltips/triggers.ts`
  * Add new triggers file to be imported by `tooltips` 
* `layout/DialogManager.tsx` 
  * Alphebetized list
 
## Functions

* `lib/positions.ts`
  * Create a `getHydratedPositionData` which mimics `orders.ts`'s `getHydratedTradeData`

## State

* `state/accountSelectors`
  * create a `getPositionDetails` hook (takes in a `marketId`) to be called by our `Triggers` dialog

## Packages

* `package.json`
  * `v4-localization`: v1.1.46 -> v1.1.50


---

<!-- Additional screenshots/recordings, before/after comparisons, testing instructions etc. -->

jaredvu commented on src/components/Collapsible.stories.tsx:
```diff
@@ -18,15 +21,15 @@ export const CollapsibleStory: Story<CollapsibleProps> = (args) => (
+CollapsibleStoryWithIconTrigger.argTypes = {
   triggerIcon: {
-    options: Object.values(IconName),
+    options: [...Object.values(IconName)],
```
qq: why does this need a shallow copy?

moo-onthelawn replied:
ah whoops, it doesn't - remenants from playing around with old edit; will remove

jaredvu commented on src/layout/DialogManager.tsx:
🙏

jaredvu commented on src/views/forms/TriggersForm/AdvancedTriggersOptions.tsx:
```diff
@@ -0,0 +1,59 @@
+
+  return (
+    <Styled.Container>
+      {
+        <Styled.Header>
+          {stringGetter({ key: STRING_KEYS.ADVANCED })}
+          {<HorizontalSeparatorFiller />}
+        </Styled.Header>
+      }
+      {/* TODO: CT-625 Update with values from abacus */}
+      {<OrderSizeInput className={className} symbol={symbol} stepSizeDecimals={stepSizeDecimals} />}
+      {<LimitPriceInputs className={className} tickSizeDecimals={tickSizeDecimals} />}
```
Are the brackets remnants of comments or conditionals?

moo-onthelawn replied:
ah sry - cleaned up - I think these are just remenants of me thinking and being bracket aggressive

jaredvu commented on src/views/forms/TriggersForm/LimitPriceInputs.tsx:
```diff
@@ -0,0 +1,78 @@
+          </WithTooltip>
+        }
+      >
+        {
+          <Styled.InputsRow>
+            <FormInput
+              id="TP-limit"
+              decimals={tickSizeDecimals || USD_DECIMALS}
+              label={
+                <>
+                  {stringGetter({ key: STRING_KEYS.TP_LIMIT })}
+                  {<Tag>USD</Tag>}
+                </>
+              }
+            />
+            <FormInput
+              id="SL-limit"
+              decimals={tickSizeDecimals || USD_DECIMALS}
+              label={
+                <>
+                  {stringGetter({ key: STRING_KEYS.SL_LIMIT })}
+                  {<Tag>USD</Tag>}
+                </>
+              }
+            />
+          </Styled.InputsRow>
+        }
```
Unnecessary brackets

jaredvu commented on src/views/forms/TriggersForm/LimitPriceInputs.tsx:
```diff
@@ -0,0 +1,78 @@
+              label={
+                <>
+                  {stringGetter({ key: STRING_KEYS.SL_LIMIT })}
+                  {<Tag>USD</Tag>}
```
nit: unnecessary brackets

jaredvu commented on src/views/forms/TriggersForm/LimitPriceInputs.tsx:
```diff
@@ -0,0 +1,78 @@
+            />
+            <FormInput
+              id="SL-limit"
+              decimals={tickSizeDecimals || USD_DECIMALS}
```
Should we use ?? instead of || to handle tickSizeDecimals = 0?

jaredvu commented on src/views/forms/TriggersForm/LimitPriceInputs.tsx:
```diff
@@ -0,0 +1,78 @@
+              label={
+                <>
+                  {stringGetter({ key: STRING_KEYS.TP_LIMIT })}
+                  {<Tag>USD</Tag>}
```
nit: unnecessary brackets

jaredvu commented on src/views/forms/TriggersForm/OrderSizeInput.tsx:
```diff
@@ -0,0 +1,67 @@
+      {/* TODO: CT-625 Update with values from abacus */}
+      <Styled.SizeInputRow>
+        <Styled.OrderSizeSlider className={className} stepSizeDecimals={stepSizeDecimals} />
+        <FormInput label="" type={InputType.Number} slotRight={<Tag>{symbol}</Tag>} />
```
blank `label`. If `<FormInput>` component is necessaray over `<Input>`, let's make label optional.

jaredvu commented on src/views/forms/TriggersForm/TriggerOrderInputs.tsx:
```diff
@@ -0,0 +1,160 @@
+        },
+      ]}
+      onValueChange={onValueChange}
+    ></DropdownSelectMenu>
```
nit: `/>` instead of empty children

jaredvu commented on src/views/forms/TriggersForm/TriggersForm.tsx:
```diff
@@ -0,0 +1,119 @@
+          }}
+          orders={takeProfitOrders}
+          onViewOrdersClick={onViewOrdersClick}
+        ></TriggerOrderInputs>
```
nit: `/>`

jaredvu commented on src/views/forms/TriggersForm/TriggersForm.tsx:
```diff
@@ -0,0 +1,119 @@
+          orders={stopLossOrders}
+          tickSizeDecimals={tickSizeDecimals}
+          onViewOrdersClick={onViewOrdersClick}
+        ></TriggerOrderInputs>
```
nit: `/>`

jaredvu commented on src/views/forms/TriggersForm/TriggersForm.tsx:
```diff
@@ -0,0 +1,119 @@
+          symbol={symbol}
+          stepSizeDecimals={stepSizeDecimals}
+          tickSizeDecimals={tickSizeDecimals}
+        ></AdvancedTriggersOptions>
```
nit: `/>`

jaredvu commented on src/views/forms/TriggersForm/TriggersForm.tsx:
```diff
@@ -0,0 +1,119 @@
+        ></AdvancedTriggersOptions>
+      }
+      <Button action={ButtonAction.Primary} state={{ isDisabled }}>
+        {isDisabled
+          ? stringGetter({ key: STRING_KEYS.ENTER_TRIGGERS })
```
I feel like there is  an intermediate step before being disabled that should control the CTA.
Like having a var `hasInputtedTriggers` that is part of the `isDisabled` check.

moo-onthelawn replied:
ah yeah - I was going to update that logic as a part of the TODO [here](https://github.com/dydxprotocol/v4-web/pull/376/files#diff-d711aaf1fda8e97dbf752a4b79c17ff1156d55a1df2174db83f0ab8e87a73ab2R34), but I can make that more clear in my variable naming for rn

moo-onthelawn commented on src/views/forms/TriggersForm/LimitPriceInputs.tsx:
```diff
@@ -0,0 +1,78 @@
+          <Styled.InputsRow>
+            <FormInput
+              id="TP-limit"
+              decimals={tickSizeDecimals || USD_DECIMALS}
```
ah makes sense!

moo-onthelawn commented on src/views/forms/TriggersForm/TriggersForm.tsx:
```diff
@@ -0,0 +1,119 @@
+        ></AdvancedTriggersOptions>
+      }
+      <Button action={ButtonAction.Primary} state={{ isDisabled }}>
+        {isDisabled
```
Possible actions are:
1. Creating a new trigger (SL/TP order) - "Add triggers"
2. Editing an existing trigger - "Edit triggers"
3. Removing a trigger order (if you clear the default input) 

*And some combination of all 3 of the above

I think there's a concern that the 3rd action isn't clear - we could use the button text here to reflect that more clearly (maybe - "Remove triggers" or "Edit and Remove triggers"). I can circle back with design on that!

moo-onthelawn commented on src/views/forms/TriggersForm/TriggersForm.tsx:
```diff
@@ -0,0 +1,119 @@
+      {
+        <TriggerOrderInputs
+          symbol={symbol}
+          tooltipId={'stop-loss'}
```
updated UI after changes:

<img width="536" alt="Screenshot 2024-03-20 at 2 55 34 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/18f69d8d-83f2-4928-b8b5-6e3b648073fe">
<img width="529" alt="Screenshot 2024-03-20 at 2 55 24 PM" src="https://github.com/dydxprotocol/v4-web/assets/70078372/4219aa47-9715-4a73-8ba6-02dfca90e798">

moo-onthelawn commented on src/views/tables/PositionsTable/PositionsTriggersCell.tsx:
```diff
@@ -132,9 +154,13 @@ export const PositionsTriggersCell = ({
+                    {stringGetter({
+                      key: isStopLossOrder(order)
+                        ? STRING_KEYS.EDIT_STOP_LOSS
+                        : STRING_KEYS.EDIT_STOP_LOSS, // TODO: CT-704 update to EDIT_TAKE_PROFIT
```
my bad here, forgot a string key 😢 made a mini task to update this later, didn't want to hold up the PR on this

moo-onthelawn commented:
updated:
- review comments
- updated "Edit [xxx]" buttons on Hovercards to call into the TriggersDialog ([here](https://github.com/dydxprotocol/v4-web/pull/376/files/418befe1d4e0cd54c31d5469c892f9e51f04a35f..970c0cb3bceb19cb82695c55971aac3dfe2aabce#diff-cc13448a26662c6c8c218ec88607fcf7cc4811e186634c217ac64a2783e56a66R110), [here](https://github.com/dydxprotocol/v4-web/pull/376/files/418befe1d4e0cd54c31d5469c892f9e51f04a35f..970c0cb3bceb19cb82695c55971aac3dfe2aabce#diff-cc13448a26662c6c8c218ec88607fcf7cc4811e186634c217ac64a2783e56a66R157))
- added [edge case](https://github.com/dydxprotocol/v4-web/pull/376/files/418befe1d4e0cd54c31d5469c892f9e51f04a35f..970c0cb3bceb19cb82695c55971aac3dfe2aabce#diff-d711aaf1fda8e97dbf752a4b79c17ff1156d55a1df2174db83f0ab8e87a73ab2R40) to hide "Advanced Section" + "Edit/Add Triggers" CTA in TriggersDialog when both SL and TP have multiple orders
